### PR TITLE
quant doc: improve rendered documentation for backend_config_dict

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#77535 quant doc: improve rendered documentation for backend_config_dict**

Summary:

This improves the documentation page for backend_config_dict to render
the configurations in a human readable format, such as

```
{
  'pattern': torch.nn.modules.pooling.AdaptiveAvgPool1d,
  'dtype_configs': [
    {
      'input_dtype': torch.quint8,
      'output_dtype': torch.quint8,
    },
    {
      'input_dtype': torch.float16,
      'weight_dtype': torch.float16,
      'bias_dtype': torch.float16,
      'output_dtype': torch.float16,
    },
  ],
  'observation_type': ObservationType.OUTPUT_SHARE_OBSERVER_WITH_INPUT,
},
```

The results are also now sorted alphabetically by the normalized name of
the root op in the pattern.

A couple of utility functions are created to help with this. If in the future
we convert backend_config_dict to use typed objects, we can move this logic
to the objects at that time.

Test plan:

```
cd docs
make html
cd build
python -m server.http
// renders correctly, example: https://gist.github.com/vkuzo/76adfc7c89e119c59813a733fa2cd56f
```

### CODE REVIEW

File: docs/source/scripts/build_quantization_configs.py [hunk 1]
```diff
@@ -20,4 +25,44 @@
...
+
+    configs.sort(key=_sort_key_func)
+
+    entries = []
+    for entry in configs:
+        entries.append(entry_to_pretty_str(entry))
+    entries = ",\n".join(entries)
+    print(entries)
+
+    sys.stdout = old_stdout
```
Comment (andrewor14) [inline docs/source/scripts/build_quantization_configs.py:68]: Is there a reason why we're not using `f.write` directly rather than assigning `sys.stdout`? This feels a bit hacky

**Review by andrewor14 — APPROVED** [review]
